### ISSUE TITLE
asserting list of sealed class instances gives type inference issue

### ISSUE THREAD

**Issue by KasperJanssens** [kotlintest/kotlintest#321]
Hello,

I have a sealed class like this
```
 sealed class TestSealed{
      data class Test1(val value : String): TestSealed()
      data class Test2(val value : Int): TestSealed()
  }
```

And I would like to assert that in a list of TestSealed my favourite one is present, like so
```
    @Test
    fun assertSealedClass () {
        val tests = listOf(TestSealed.Test1("test1"),
                TestSealed.Test2(2))

        tests should contain(TestSealed.Test2(2))
    }
```

However, I get a type inference failure. When I create a list that is only populated with one of the ADTs, like only Test1 or only Test2, the should contain works. Otherwise it does not. I was wondering whether this is expected or if there is another way to achieve this (did not get lucky with type annotations). I can use tests.contain(...) shouldBe true of course.

**Comment by sksamuel** [comment]
A `Matcher<T>` can be combined with another matcher, via `and` or `or` and so a `Matcher<T>` cannot be variant, which would allow a `Matcher<T>` to be used when we want a Matcher<U> for `U  <: T.` So I cannot redefine the Matcher type.

But it's possible to change this particular matcher to return a C: Collection<T> would would allow the type inference to infer List<T> in this case, and then it works. I'll do this for 3.1.0

**Comment by KasperJanssens** [comment]
Okay, I'll keep this open then until the 3.1 is done or you prefer that I close it?

**Comment by sksamuel** [comment]
I'll close it when I do the commit later.

**Comment by sksamuel** [comment]
This is now fixed in release 3.1.0-RC4.
